Generate code that is friendlier to rustc's item_bodies_checking

dotdash opened the pull request:
Having the Some() constructor in each match arm body puts extras stress
on rustc's item_bodies_checking pass. We can work around that by moving
the Some() constructor around the match, and directly returning None
from the default arm, which is the only one that generates a None value.

On my box, this almost cuts the time spent in item_bodies_checking in
half, going from about 8.7s to about 4.6s, and reduces the complete
compile time from about 13s to about 9s, so about a third less.

There are no changes in performance in `cargo bench`.

Note that doing the same for the composition_table() function does not
yield any compile time wins, but would cause a performance regression.

cc #29

sujayakar commented on scripts/unicode.py:
```diff
@@ -315,14 +315,14 @@ def gen_decomposition_tables(canon_decomp, compat_decomp, out):
         out.write("#[inline]\n")
         out.write("pub fn %s_fully_decomposed(c: char) -> Option<&'static [char]> {\n" % name)
-        out.write("    match c {\n")
+        out.write("    Some(match c {\n")
```
@dotdash, can you add a comment here explaining the motivation for wrapping the match in `Some`?  otherwise, it may look non-idiomatic to future readers of this codegen script.

Manishearth commented:
r=me with @sujayakar's comment addressed

dotdash commented:
Added the requested comment, and adjusted the commit message to say that there is no change in performance, rather than in the generated assembly code, as `--emit asm` doesn't actually show that, because all the relevant functions are marked as `#[inline]`, so there's only MIR generated for them.
